### ISSUE TITLE
unable to reach the webui

### ISSUE THREAD

**Issue by jlim0930** [MarkusMcNugen/docker-qBittorrentvpn#27]
my docker run line
/usr/bin/docker run \
        --name qbitvpn \
        --privileged \
        -p 8080:8080 \
        -p 6882:6882  \
        -p 6882:6882/udp \
        -e PUID=1026 \
        -e PGID=100 \
        -e "VPN_ENABLED=yes" \
        -e "NAME_SERVERS=8.8.8.8,1.1.1.1" \
        -e "LAN_NETWORK=192.168.1.0/24" \
        -v /data/config/qbitvpn:/config \
        -v /data/downloads/qbtvpn:/downloads \
        --log-driver=journald \
        --log-opt tag="|{{.Name}}|{{.ID}}|" \
        markusmcnugen/qbittorrentvpn

and i can see that its running and right ports are being exposed and forwarded
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS                  PORTS                                                                                        NAMES
e5f84be326bc        markusmcnugen/qbittorrentvpn   "/bin/bash /etc/open…"   1 second ago        Up Less than a second   0.0.0.0:6882->6882/tcp, 0.0.0.0:8080->8080/tcp, 0.0.0.0:6882->6882/udp, 8999/tcp, 8999/udp   qbitvpn

however something is keep disconnecting me from port 8080 and unable to get to the webui
$ telnet localhost 8080
Trying ::1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.

**Comment by NathanHRich** [comment]
Any updates on this? I'm having the same issue.
EDIT Here's my docker run line:
```
docker run --privileged  -d \
--name=qbittorrent \
--restart=always \
--log-opt max-size=50m \
-v /docker/containers/qbittorrent:/config \
-v /plex/downloads:/downloads \
-e "VPN_ENABLED=yes" \
-e "LAN_NETWORK=192.168.86.0/24" \
-e "NAME_SERVERS=8.8.8.8,8.8.4.4" \
-e "VPN_USERNAME=##########" \
-e "VPN_PASSWORD=##########" \
-e PGID=1000 -e PUID=1000 \
-p 8080:8080 \
-p 8999:8999 \
-p 8999:8999/udp \
markusmcnugen/qbittorrentvpn
```

**Comment by themfinger** [comment]
I think I also have everything well configured; I can read in see the log that everything looks nice with a line finishing:

 [info] Started qBittorrent daemon successfully...
And during the launch, it shows a the following in the log:

    _"[info] WebUI port defined as"_

I set an environment variable  **"WEBUI_PORT_ENV=8099"**, which is correctly bridged.

**But, when I try to connect, nothing happens.**

After reading other issues, it seems the VPN module does't configure the defined WEBUI Port:
_(N) 2019-08-12T23:20:08 - Using built-in Web UI.
(N) 2019-08-12T23:20:08 - Web UI translation for selected locale (en) has been successfully loaded.
(N) 2019-08-12T23:20:08 - Web UI: Now listening on IP: *, port: 8080
(I) 2019-08-12T23:20:08 - System network status changed to ONLINE
(I) 2019-08-12T23:20:08 - qBittorrent is successfully listening on interface :: port: TCP/8999
(I) 2019-08-12T23:20:08 - qBittorrent is successfully listening on interface 0.0.0.0 port: TCP/8999
(I) 2019-08-12T23:20:08 - qBittorrent is successfully listening on interface 0.0.0.0 port: UDP/8999_

IPTABLES:
_2019-08-12 23:20:07.971489 [info] iptables defined as follows...
-A INPUT -i eth0 -p tcp -m tcp --dport 8080 -j ACCEPT
-A INPUT -i eth0 -p tcp -m tcp --sport 8080 -j ACCEPT
-A OUTPUT -o eth0 -p tcp -m tcp --dport 8080 -j ACCEPT_

Can someone please help?
Thanks in advance,

**Comment by michecode** [comment]
hey the documentation is wrong. i accidentally closed my issue ( #26 ). instead of WEBUI_PORT_ENV you should use WEBUI_PORT

**Comment by norey** [comment]
I'm unable to reach the WebUI as well. This is how I am running it:

```
sudo docker run --privileged  -d \
              -v /path/config/:/config \
              -v /path/downloads/:/downloads \
              -e "VPN_ENABLED=yes" \
              -e "LAN_NETWORK=192.168.1.0/24" \
              -e "NAME_SERVERS=8.8.8.8,8.8.4.4" \
              -e "VPN_USERNAME=###" \
              -e "VPN_PASSWORD=###" \
              -e "WEBUI_PORT=8080" \
              -e PUID=1001 \
              -e PGID=1001 \
              -e UMASK=002 \
              -p 9090:8080 \
              -p 9999:8999 \
              -p 9999:8999/udp \
              --name qbt \
              markusmcnugen/qbittorrentvpn
```

Some logs, well important parts of those

```
$ sudo docker logs qbt

Sun Oct 13 19:03:30 2019 Initialization Sequence Completed
2019-10-13 19:03:32.167770 [info] WebUI port defined as 8080
2019-10-13 19:03:32.244310 [info] LAN Network defined as 192.168.1.0/24
2019-10-13 19:03:32.320642 [info] Default gateway defined as 172.17.0.1
2019-10-13 19:03:32.394927 [info] ip route defined as follows...
--------------------
...
...
--------------------
2019-10-13 19:03:32.513336 [info] Docker network defined as 172.17.0.0/16
2019-10-13 19:03:32.706395 [info] iptables defined as follows...
--------------------
-P INPUT DROP
-P FORWARD ACCEPT
-P OUTPUT DROP
-A INPUT -i tun0 -j ACCEPT
-A INPUT -s 172.17.0.0/16 -d 172.17.0.0/16 -j ACCEPT
-A INPUT -i eth0 -p udp -m udp --sport 1194 -j ACCEPT
-A INPUT -i eth0 -p tcp -m tcp --dport 8080 -j ACCEPT
-A INPUT -i eth0 -p tcp -m tcp --sport 8080 -j ACCEPT
-A INPUT -s 192.168.1.0/24 -i eth0 -p tcp -m tcp --dport 8999 -j ACCEPT
-A INPUT -p icmp -m icmp --icmp-type 0 -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A OUTPUT -o tun0 -j ACCEPT
-A OUTPUT -s 172.17.0.0/16 -d 172.17.0.0/16 -j ACCEPT
-A OUTPUT -o eth0 -p udp -m udp --dport 1194 -j ACCEPT
-A OUTPUT -o eth0 -p tcp -m tcp --dport 8080 -j ACCEPT
-A OUTPUT -o eth0 -p tcp -m tcp --sport 8080 -j ACCEPT
-A OUTPUT -d 192.168.1.0/24 -o eth0 -p tcp -m tcp --sport 8999 -j ACCEPT
-A OUTPUT -p icmp -m icmp --icmp-type 8 -j ACCEPT
-A OUTPUT -o lo -j ACCEPT
--------------------
Adding 1001 group
Adding 1001 user
2019-10-13 19:03:32.982605 [info] UMASK defined as '002'
2019-10-13 19:03:33.068209 [info] Starting qBittorrent daemon...
Logging to /config/qBittorrent/data/logs/qbittorrent-daemon.log.
2019-10-13 19:03:34.150376 [info] qBittorrent PID: 192
2019-10-13 19:03:34.169745 [info] Started qBittorrent daemon successfully...


$ sudo docker exec qbt cat /config/qBittorrent/data/logs/qbittorrent.log

...
(W) 2019-10-13T05:31:01 - WebUI: Invalid Host header, port mismatch. Request source IP: '::ffff:192.168.1.10'. Server port: '8080'. Received Host header: 'homesv:9090'
(W) 2019-10-13T05:31:02 - WebUI: Invalid Host header, port mismatch. Request source IP: '::ffff:192.168.1.10'. Server port: '8080'. Received Host header: 'homesv:9090'
(W) 2019-10-13T05:31:02 - WebUI: Invalid Host header, port mismatch. Request source IP: '::ffff:192.168.1.10'. Server port: '8080'. Received Host header: 'homesv:9090'
...
```

I'm really not sure if this the right usage of "WEBUI_PORT" or if there is a need to use it since I changed the ports to 9090/9999 locally since other services are listening on the default ones.

**Comment by eepohboy** [comment]
I'm having this problem as well running on Synology Docker. Below are my logs. The worrying output I get is the one that says:

`RTNETLINK answers: Network is unreachable`

When I turn set VPN_ENABLED to 'no'. I can access the WEBUI on 9083 without any problems. 

I also notice when I disable the VPN, the qBittorrent logs get an External IP. I don't get this when the VPN is running

I would like to get this running with a VPN if possible. Any help would be appreciated.

Docker logs:
`
2019-10-16 07:15:25.856827 [info] VPN_ENABLED defined as 'yes'
2019-10-16 07:15:26.048342 [info] OpenVPN config file (ovpn extension) is located at /config/openvpn/gw1.sin1.slickvpn.com.ovpn
dos2unix: converting file /config/openvpn/gw1.sin1.slickvpn.com.ovpn to Unix format...
2019-10-16 07:15:26.181520 [info] VPN remote line defined as 'gw1.sin1.slickvpn.com 443 udp'
2019-10-16 07:15:26.239917 [info] VPN_REMOTE defined as 'gw1.sin1.slickvpn.com'
2019-10-16 07:15:26.297776 [info] VPN_PORT defined as '443'
2019-10-16 07:15:26.357139 [info] VPN_PROTOCOL defined as 'udp'
2019-10-16 07:15:26.416117 [info] VPN_DEVICE_TYPE defined as 'tun0'
2019-10-16 07:15:26.473871 [info] LAN_NETWORK defined as '10.0.2.0/24'
2019-10-16 07:15:26.541066 [info] NAME_SERVERS defined as '1.1.1.1,1.0.0.1'
2019-10-16 07:15:26.598615 [info] VPN_OPTIONS not defined (via -e VPN_OPTIONS)
2019-10-16 07:15:26.657698 [info] Adding 1.1.1.1 to resolv.conf
2019-10-16 07:15:26.716317 [info] Adding 1.0.0.1 to resolv.conf
2019-10-16 07:15:26.791778 [info] Starting OpenVPN...
Wed Oct 16 07:15:26 2019 WARNING: file 'credentials.conf' is group or others accessible
Wed Oct 16 07:15:26 2019 OpenVPN 2.4.4 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Sep  5 2018
Wed Oct 16 07:15:26 2019 library versions: OpenSSL 1.1.0g  2 Nov 2017, LZO 2.08
Wed Oct 16 07:15:26 2019 TCP/UDP: Preserving recently used remote address: [AF_INET]199.195.193.181:443
Wed Oct 16 07:15:26 2019 UDP link local: (not bound)
Wed Oct 16 07:15:26 2019 UDP link remote: [AF_INET]199.195.193.181:443
Wed Oct 16 07:15:26 2019 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
Wed Oct 16 07:15:27 2019 [VPN] Peer Connection Initiated with [AF_INET]199.195.193.181:443
Wed Oct 16 07:15:28 2019 TUN/TAP device tun0 opened
Wed Oct 16 07:15:28 2019 do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Wed Oct 16 07:15:28 2019 /sbin/ip link set dev tun0 up mtu 1500
Wed Oct 16 07:15:28 2019 /sbin/ip addr add dev tun0 local 10.10.8.22 peer 10.10.8.21
Wed Oct 16 07:15:28 2019 Initialization Sequence Completed
2019-10-16 07:15:31.863722 [info] WebUI port defined as 9083
2019-10-16 07:15:31.921465 [info] LAN Network defined as 10.0.2.0/24
2019-10-16 07:15:31.979063 [info] Default gateway defined as 10.10.8.21
RTNETLINK answers: Network is unreachable
2019-10-16 07:15:32.035453 [info] ip route defined as follows...
--------------------
default via 10.10.8.21 dev tun0 
10.10.8.1 via 10.10.8.21 dev tun0 
10.10.8.21 dev tun0 proto kernel scope link src 10.10.8.22 
172.17.0.0/16 dev eth0 proto kernel scope link src 172.17.0.6 
199.195.193.181 via 172.17.0.1 dev eth0 
--------------------
iptable_mangle          1656  0
ip_tables              14092  3 iptable_filter,iptable_mangle,iptable_nat
x_tables               16302  19 ip6table_filter,xt_ipvs,xt_iprange,xt_mark,xt_recent,ip_tables,xt_tcpudp,ipt_MASQUERADE,xt_limit,xt_state,xt_conntrack,xt_LOG,xt_nat,xt_multiport,iptable_filter,xt_REDIRECT,iptable_mangle,ip6_tables,xt_addrtype
2019-10-16 07:15:32.099917 [info] iptable_mangle support detected, adding fwmark for tables
2019-10-16 07:15:32.187223 [info] Docker network defined as 172.17.0.0/16
2019-10-16 07:15:32.343543 [info] iptables defined as follows...
--------------------
-P INPUT DROP
-P FORWARD ACCEPT
-P OUTPUT DROP
-A INPUT -i tun0 -j ACCEPT
-A INPUT -s 172.17.0.0/16 -d 172.17.0.0/16 -j ACCEPT
-A INPUT -i eth0 -p udp -m udp --sport 443 -j ACCEPT
-A INPUT -i eth0 -p tcp -m tcp --dport 9083 -j ACCEPT
-A INPUT -i eth0 -p tcp -m tcp --sport 9083 -j ACCEPT
-A INPUT -s 10.0.2.0/24 -i eth0 -p tcp -m tcp --dport 8999 -j ACCEPT
-A INPUT -p icmp -m icmp --icmp-type 0 -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A OUTPUT -o tun0 -j ACCEPT
-A OUTPUT -s 172.17.0.0/16 -d 172.17.0.0/16 -j ACCEPT
-A OUTPUT -o eth0 -p udp -m udp --dport 443 -j ACCEPT
-A OUTPUT -o eth0 -p tcp -m tcp --dport 9083 -j ACCEPT
-A OUTPUT -o eth0 -p tcp -m tcp --sport 9083 -j ACCEPT
-A OUTPUT -d 10.0.2.0/24 -o eth0 -p tcp -m tcp --sport 8999 -j ACCEPT
-A OUTPUT -p icmp -m icmp --icmp-type 8 -j ACCEPT
-A OUTPUT -o lo -j ACCEPT
--------------------
Adding 100 group
groupadd: GID '100' already exists
Adding 1026 user
useradd: user 'qbittorent' already exists
2019-10-16 07:15:32.454826 [warn] UMASK not defined (via -e UMASK), defaulting to '002'
2019-10-16 07:15:32.512446 [info] Starting qBittorrent daemon...
Logging to /config/qBittorrent/data/logs/qbittorrent-daemon.log.
2019-10-16 07:15:33.578624 [info] qBittorrent PID: 190
2019-10-16 07:15:33.596591 [info] Started qBittorrent daemon successfully...
`

qBittorrent logs:
`
(N) 2019-10-16T07:15:32 - qBittorrent v4.1.5 started
(I) 2019-10-16T07:15:32 - qBittorrent is trying to listen on any interface port: 8999
(N) 2019-10-16T07:15:32 - Peer ID: -qB4150-
(N) 2019-10-16T07:15:32 - HTTP User-Agent is 'qBittorrent/4.1.5'
(I) 2019-10-16T07:15:32 - DHT support [ON]
(I) 2019-10-16T07:15:32 - Local Peer Discovery support [ON]
(I) 2019-10-16T07:15:32 - PeX support [ON]
(I) 2019-10-16T07:15:32 - Anonymous mode [OFF]
(I) 2019-10-16T07:15:32 - Encryption support [ON]
(I) 2019-10-16T07:15:32 - Embedded Tracker [OFF]
(I) 2019-10-16T07:15:32 - GeoIP database loaded. Type: GeoLite2-Country. Build time: Tue Oct 8 19:01:04 2019.
(N) 2019-10-16T07:15:32 - Using built-in Web UI.
(N) 2019-10-16T07:15:32 - Web UI translation for selected locale (en) has been successfully loaded.
(N) 2019-10-16T07:15:32 - Web UI: Now listening on IP: *, port: 9083
(I) 2019-10-16T07:15:32 - qBittorrent is successfully listening on interface :: port: TCP/8999
(I) 2019-10-16T07:15:32 - qBittorrent is successfully listening on interface 0.0.0.0 port: TCP/8999
(I) 2019-10-16T07:15:32 - qBittorrent is successfully listening on interface 0.0.0.0 port: UDP/8999
`

**Comment by eepohboy** [comment]
I am having the same issue when running the container on my Synology box. What worries me is the log output that says

`RTNETLINK answers: Network is unreachable`

Also, I notice I can access the WEBUI when I turn off the VPN by setting VPN_ENABLED to 'no'. When I do this I get an External IP set in the qBittorrent logs, but with the VPN on, this last log entry does not appear

Docker logs
[qbittorrentvpn.docker.log](https://github.com/MarkusMcNugen/docker-qBittorrentvpn/files/3733134/qbittorrentvpn.docker.log)

qBittorrent logs
[qbittorrent.log](https://github.com/MarkusMcNugen/docker-qBittorrentvpn/files/3733135/qbittorrent.log)

**Comment by simepy** [comment]
You can try to enable module "iptable_magle" with "modprod" in your host
with `/sbin/modprobe iptable_mangle`

For me it's work :)

**Comment by eepohboy** [comment]
> You can try to enable module "iptable_magle" with "modprod" in your host
> with `/sbin/modprobe iptable_mangle`

Does this affect the security of the host in any way?

**Comment by Altycoder** [comment]
> hey the documentation is wrong. i accidentally closed my issue ( #26 ). instead of WEBUI_PORT_ENV you should use WEBUI_PORT

I have tried this but it doesn't work so I've just raised a new issue

**Comment by norey** [comment]
any updates on this? anyone able to run the docker image correctly and get access to the WebUI?

**Comment by scb147** [comment]
Since whatever the last update is, I cannot get the WebUI to work at all. It was working fine with my WEBUI_PORT and WEBUI_PORT_ENV set to 6060, with the VPN enabled.  Now I can't even get it to work with the VPN disabled.

**Comment by aes-grave** [comment]
I am having a similar issue, but I think it's actually something to do with a race condition. If you netstat -lntu, do you have an entry like 

tcp 0.0.0.0:8080 

I don't, I only have

`tcp        0      0 0.0.0.0:8999            0.0.0.0:*               LISTEN     
tcp        0      0 127.0.0.11:37791        0.0.0.0:*               LISTEN     
tcp6       0      0 :::8080                 :::*                    LISTEN    `

I think my problem and it could be some of yours as well is I think it's not mapping port 8080 for tcp only for tcp6. Which is blocked by the iptable rules.

**Comment by nimblebytes** [comment]
There are two issues at play here. As @sarkyscouser stated
 
> > hey the documentation is wrong. i accidentally closed my issue ( #26 ). instead of WEBUI_PORT_ENV you should use WEBUI_PORT
> 
> I have tried this but it doesn't work so I've just raised a new issue

Secondly, there is an error in the code that messed up the port. Running the container a second time with the same config arguments "fixes" it. See https://github.com/MarkusMcNugen/docker-qBittorrentvpn/issues/33#issuecomment-585420773

**Comment by isudoi** [comment]
I have Synology NAS Docker Setup
PIA is my VPN Provider (Mullvad was my previous but it was giving heck of hard time to get anything connected)

So, I was able to get it working, kinda, and the container is running and connected via VPN now


--------------------


```
Adding 0 group
groupadd: GID '0' already exists
Adding 0 user
useradd: UID 0 is not unique
2021-02-07 20:01:50.246901 [warn] UMASK not defined (via -e UMASK), defaulting to '002'
2021-02-07 20:01:50.303396 [info] Starting qBittorrent daemon...
Logging to /config/qBittorrent/data/logs/qbittorrent-daemon.log.
2021-02-07 20:01:51.366776 [info] qBittorrent PID: 185
2021-02-07 20:01:51.384442 [info] Started qBittorrent daemon successfully...
```



And I was able to shell to the container and run ifconfig to confirm the public IP address is the VPN one, which is great.

```
root@Synology:~# docker exec -it 4b2377436d22 /bin/sh
# curl ifconfig.io
154.3.42.12
```

However, still UI is not loading



My docker stack is:

```
---
version: "2"
services:
  qbittorrentvpn:
    image: markusmcnugen/qbittorrentvpn
    container_name: qbittorrentvpn
    privileged: true   
    environment:
      - VPN_USERNAME=username
      - VPN_PASSWORD=password
      - PUID=0 #optional
      - PGID=0 #optional
      - WEBUI_PORT_ENV=8787 #optional
      - INCOMING_PORT_ENV=8999 #optional
      - VPN_ENABLED=yes
      - LAN_NETWORK=10.0.12.0/24
      - NAME_SERVERS=8.8.8.8,1.1.1.1
    ports:
      - 8787:8787
      - 8999:8999
      - 8999:8999/udp
    volumes:
      - /volume1/docker/qBittorrentVPN/config:/config
      - /volume1/docker/qBittorrentVPN/downloads:/downloads
    restart: unless-stopped
```





Noticed, one thing, that parameter that needs to be changed

```
      - WEBUI_PORT_ENV=8787 #optional
      - INCOMING_PORT_ENV=8999 #optional
```

Use instead this:

```
      - WEBUI_PORT=8787 #optional
      - INCOMING_PORT=8999 #optional
```

so when I have update the stack 

ran again:

```
root@Synology:~# docker exec -it de597f9d4765 /bin/sh
# curl ifconfig.io
66.115.142.72

```


and it worked like charm

**Comment by triksmelb** [comment]
no luck accessing the webui otherwise the container seems to run.
1. tried via SSH docker run 
2. below in portainer > stack
3. ran the command "docker exec -it ""containername"" /bin/sh" which returned "#"
Strangely transmissionopenvpn is working fine.

version: "2"
services:
  qbittorrentvpn:
    image: markusmcnugen/qbittorrentvpn
    container_name: qbittorrentvpn
    privileged: true   
    environment:
      - VPN_USERNAME=######
      - VPN_PASSWORD=######
      - PUID=1026
      - PGID=100
      - WEBUI_PORT=8787
      - INCOMING_PORT=8999
      - VPN_ENABLED=yes
      - LAN_NETWORK=10.0.1.0/24
      - NAME_SERVERS=1.1.1.1;8.8.8.8
    ports:
      - 8787:8787
      - 8999:8999
      - 8999:8999/udp
    volumes:
      - /volume1/docker/qbittorrentvpn:/config
      - /volume1/downloads/qbittorrent/:/downloads
    restart: unless-stopped

**Comment by iwindon** [comment]
I seem to have the same issue.  I've not been able to resolve the issue as of yet. Attached are my logs:
[qbittorrentvpn_logs-5.txt](https://github.com/MarkusMcNugen/docker-qBittorrentvpn/files/10273233/qbittorrentvpn_logs-5.txt)

and my yml in the stack from portainer running on docker

qbittorrentvpn:
    image: markusmcnugen/qbittorrentvpn
    container_name: qbittorrentvpn
    privileged: true
    environment:
      - VPN_USERNAME=xxxx
      - VPN_PASSWORD=xxxx
      - PUID=998
      - PGID=100
      - WEBUI_PORT=8080
      - INCOMING_PORT=8999
      - VPN_ENABLED=yes
      - LAN_NETWORK=192.168.1.0/24 #adjust this to YOUR network settings
      - NAME_SERVERS=8.8.8.8,8.8.4.4 #you can use whatever DNS provider you want
    ports:
      - 8080:8080
      - 8999:8999
      - 8999:8999/udp
    volumes:
      - /QBittorrentVPN:/config
      - /mediaserver/Torrents:/downloads
      - /plex/media:/media
    restart: unless-stopped

docker version is:

[root@saturn openvpn]# docker --version
Docker version 20.10.22, build 3a2c30b

Running on RHEL 9.1

**Comment by noob34433** [comment]
any updates? having the same issue....

**Comment by LeqitSebi** [comment]
> You can try to enable module "iptable_magle" with "modprod" in your host with `/sbin/modprobe iptable_mangle`
> 
> For me it's work :)

After days of trying, this is the solution that works for me as well!

**Comment by tejaschokhawala** [comment]
> > You can try to enable module "iptable_magle" with "modprod" in your host with `/sbin/modprobe iptable_mangle`
> > For me it's work :)
> 
> After days of trying, this is the solution that works for me as well!

For me too. been trying for whole day now
